File: clsDB.cs
using System.Data;
using System.Windows.Forms;
using MbcsCentral;

public class clsDB
{
    LogClass Log = LogClass.instance;

	protected const string DS_DATA_SET = "wintrm";
	protected string loadStmt;
	protected string loadAllStmt;
	protected string insertStmt;
	protected string updateStmt;
	protected string localTable;
	protected string priKeyName;
	protected System.Data.SqlClient.SqlConnection localCN;

    protected string connectionString;
	
	protected string SQL_DELETE; // VBConversions Note: Initial value of ""DELETE FROM " + localTable + " WHERE " + priKeyName + " = ?;"" cannot be assigned here since it is non-static.  Assignment has been moved to the class constructors.

    public clsDB() { }

	public clsDB(string table, string priKeyName, string loadStmt, string loadAllStmt, string insertStmt, string updateStmt, System.Data.SqlClient.SqlConnection cn)
	{
		// VBConversions Note: Non-static class variable initialization is below.  Class variables cannot be initially assigned non-static values in C#.
		SQL_DELETE = "DELETE FROM " + localTable + " WHERE " + priKeyName + " = ?;";
		
		
		try
		{
			this.localTable = table;
			this.priKeyName = priKeyName;
			this.loadStmt = loadStmt;
			this.loadAllStmt = loadAllStmt;
			this.insertStmt = insertStmt;
			this.updateStmt = updateStmt;
			//this.localCN = cn;
            this.connectionString = cn.ConnectionString;
			SQL_DELETE = "DELETE FROM " + localTable + " WHERE [" + priKeyName + "] = ?;";
		}
		catch (System.Exception)
		{
			return;
		}
	}
	
	protected static string prepareStatement(string sql, string[] options)
	{
		string s = "";
		int i;
		int n = 0;
		
		if (options == null || (options.Length - 1) < 0)
		{
			s = sql;
		}
		else
		{
			for (i = 1; i <= sql.Length; i++)
			{
				if (sql.Substring(i - 1, 1) == "?")
				{
					if (n <= (options.Length - 1))
					{
						if (! (options[n] == null))
						{
							if (options[n].ToLower().Equals("true"))
							{
								options[n] = "1";
							}
							else if (options[n].ToLower().Equals("false"))
							{
								options[n] = "0";
							}
						}
						s += options[n];
						n++;
					}
					else
					{
						
					}
				}
				else
				{
					s += sql.Substring(i - 1, 1);
				}
			}
		}
		return s;
		
	}
	
	protected DataSet loadByPrimaryKey(string key)
	{
		
		string[] options = new string[2];
		options[0] = key;
		try
		{
			//localCN.Open();
            //localCN.ConnectionString = connectionString;
			string stmt = prepareStatement(loadStmt, options);
			System.Data.SqlClient.SqlDataAdapter da = new System.Data.SqlClient.SqlDataAdapter(stmt, this.connectionString);
			DataSet ds = new DataSet(DS_DATA_SET);
			da.Fill(ds, DS_DATA_SET);
			
			return ds;
		}
		catch (System.Data.SqlClient.SqlException ex)
		{
			MessageBox.Show("Error while retrieving record from " + localTable + ": " + ex.Message);
			return null;
		}
		catch (System.Exception ex)
		{
			Log.Log((int)LogClass.logType.ErrorCondition, "Error while retrieveing record from " + localTable + ": " + ex.Message);
			return null;
		}
		finally
		{
			try
			{
				localCN.Close();
			}
			catch
			{
			}
		}
		
	}
	
	protected DataSet loadAll()
	{
		
		/* if (localCN.State != ConnectionState.Open)
		{
			try
			{
				localCN.Open();
			}
			catch (System.Data.SqlClient.SqlException ex)
			{
				Log.Log((int)LogClass.logType.ErrorCondition, "Opening database: <" + localCN.ConnectionString + ">" + "\r\n" + "   " + loadAllStmt + "\r\n" + "   " + ex.Message, false, MessageBoxButtons.OK);
				return null;
			}
			catch (System.Exception ex)
			{
				Log.Log((int)LogClass.logType.ErrorCondition, "Opening database: <" + localCN.ConnectionString + ">" + "\r\n" + "   " + loadAllStmt + "\r\n" + "   " + ex.Message);
				return null;
			}
		}   */
		try
		{
            //localCN.ConnectionString = connectionString;
            System.Data.SqlClient.SqlDataAdapter da = new System.Data.SqlClient.SqlDataAdapter(loadAllStmt, this.connectionString);
			DataSet ds = new DataSet(DS_DATA_SET);
			da.Fill(ds, DS_DATA_SET);
			return ds;
		}
		catch (System.Data.SqlClient.SqlException ex)
		{
			Log.Log((int)LogClass.logType.ErrorCondition, "Error while retrieving records from " + localTable + ": " + ex.Message, false, MessageBoxButtons.OK);
			return null;
		}
		catch (System.Exception ex)
		{
			Log.Log((int)LogClass.logType.ErrorCondition, "Error while retrieving records from " + localTable + ": " + ex.Message, false, MessageBoxButtons.OK);
		}
		finally
		{
			try
			{
				
				localCN.Close();
			}
			catch
			{
			}
		}
		return null;
		
	}
	
	protected DataSet load(string stmt, string[] options)
	{
		
		if (options != null)
		{
			if (options.Length > 0 && options[0].Trim() != "")
			{
				stmt = prepareStatement(stmt, options);
			}
		}
        /*
		if (localCN.State != ConnectionState.Open)
		{
			try
			{
				localCN.Open();
			}
			catch (System.Exception)
			{
				return null;
			}
		}
         */
        System.Data.SqlClient.SqlDataAdapter da = new System.Data.SqlClient.SqlDataAdapter(stmt, this.connectionString);
		DataSet ds = new DataSet(DS_DATA_SET);
		try
		{
			da.Fill(ds, DS_DATA_SET);
            /*try
            {
                localCN.Close();
            }
            catch
            {
            } */
        }
		catch (System.Data.SqlClient.SqlException)
		{
			return null;
		} 
		return ds;
		
	}
	
	protected bool updateRecord(string key, string[] values)
	{
		try
        {
            try
            {
                localCN = new System.Data.SqlClient.SqlConnection(connectionString);
                localCN.Open();
            }
            catch { }  
            System.Data.SqlClient.SqlCommand cmd = new System.Data.SqlClient.SqlCommand();
			cmd.CommandType = CommandType.Text;
			cmd.Connection = localCN;
			cmd.CommandText = prepareStatement(this.updateStmt, values);
			cmd.ExecuteNonQuery();
		}
		catch (System.Data.SqlClient.SqlException ex)
		{
			MessageBox.Show("Error while updating record: " + ex.Message);
			return false;
		}
		finally
		{
			try
			{
				localCN.Close();
			}
			catch
			{
			}
		}
		return true;
	}
	
	protected int insertRecord(string[] values)
	{
		int id = -1;
		try
		{
            try
            {
                localCN = new System.Data.SqlClient.SqlConnection( connectionString);
                localCN.Open();
            }
            catch (System.Exception ex)
            {
            	
            }
            System.Data.SqlClient.SqlCommand cmd = new System.Data.SqlClient.SqlCommand();
			cmd.CommandType = CommandType.Text;
			cmd.Connection = localCN;
			cmd.CommandText = prepareStatement(this.insertStmt, values);
			cmd.ExecuteNonQuery();
			cmd.CommandText = "SELECT @@Identity";
			id = System.Convert.ToInt32(cmd.ExecuteScalar());
			//id = 0
		}
        catch (System.Data.SqlClient.SqlException ex)
		{
			if (ex.Message.ToLower().IndexOf("duplicate") + 1 > 0)
			{
				MessageBox.Show("Attempt to add a duplicate record.");
			}
			else
			{
				MessageBox.Show("Error while inserting record into " + localTable + ": " + ex.Message);
			}
			return id;
		}
		finally
		{
			try
			{
				localCN.Close();
			}
			catch
			{
			}
		}
		return id;
	}
	
	public bool delete(string key)
	{
		
		try
		{
            localCN = new System.Data.SqlClient.SqlConnection( connectionString);
            localCN.Open();
			string[] values = new string[] {key};
            System.Data.SqlClient.SqlCommand cmd = new System.Data.SqlClient.SqlCommand();
			cmd.CommandType = CommandType.Text;
			cmd.Connection = localCN;
			cmd.CommandText = prepareStatement(SQL_DELETE, values);
			cmd.ExecuteNonQuery();
		}
        catch (System.Data.SqlClient.SqlException ex)
		{
			MessageBox.Show("Error while deleting record: " + ex.Message);
			return false;
		}
		finally
		{
			try
			{
				localCN.Close();
			}
			catch
			{
			}
		}
		return true;
		
	}

    protected bool query(string stmt, string[] options)
    {

        try
        {
            localCN = new System.Data.SqlClient.SqlConnection( connectionString);
            localCN.Open();
        }
        catch (System.Exception ex)
        {
            GlobalShared.Log.Log((int)LogClass.logType.ErrorCondition, "Unable to open DB connection <" + localCN.ConnectionString + "> for query <" + stmt + ">: " + ex.Message);
            return false;
        }
        try
        {
            System.Data.SqlClient.SqlCommand cmd = new System.Data.SqlClient.SqlCommand();
            cmd.CommandType = CommandType.Text;
            cmd.Connection = localCN;
            cmd.CommandText = prepareStatement(stmt, options);
            cmd.ExecuteNonQuery();
        }
        catch (System.Data.SqlClient.SqlException ex)
        {
            GlobalShared.Log.Log((int)LogClass.logType.ErrorCondition, "Executing Query <" + stmt + ">: " + ex.Message);
            return false;
        }
        finally
        {
            try
            {
                localCN.Close();
            }
            catch { }
        }
 
        return true;

    }

    public static DataSet Query(string stmt, System.Data.SqlClient.SqlConnection localCN)
    {

        System.Data.SqlClient.SqlDataAdapter da = new System.Data.SqlClient.SqlDataAdapter(stmt, localCN.ConnectionString);
        DataSet ds = new DataSet(DS_DATA_SET);
        try
        {
            da.Fill(ds, DS_DATA_SET);
            try
            {
                localCN.Close();
            }
            catch
            {
            }
        }
        catch (System.Data.SqlClient.SqlException)
        {
            return null;
        }
        return ds;

    }

    public System.Data.SqlClient.SqlConnection LocalCN
    {
        get { return localCN; }
        set { localCN = value;
        this.connectionString = LocalCN.ConnectionString;
        }
    }
}

File: clsDB.vb
Public Class clsDB

    Protected Const DS_DATA_SET = "LeaseLetters"
    Protected loadStmt As String
    Protected loadAllStmt As String
    Protected insertStmt As String
    Protected updateStmt As String
    Protected localTable As String
    Protected priKeyName As String
    Protected created As Boolean

    Protected SQL_DELETE As String = "DELETE FROM " & _
                                    localTable & _
                                    " WHERE " & priKeyName & _
                                    " = ?;"

    Public Sub New(ByVal table As String, ByVal priKeyName As String, ByVal loadStmt As String, ByVal loadAllStmt As String, ByVal insertStmt As String, ByVal updateStmt As String)

        Me.localTable = table
        Me.priKeyName = priKeyName
        Me.loadStmt = loadStmt
        Me.loadAllStmt = loadAllStmt
        Me.insertStmt = insertStmt
        Me.updateStmt = updateStmt
        SQL_DELETE = "DELETE FROM " & _
                     localTable & _
                     " WHERE " & priKeyName & _
                     " = ?;"
    End Sub

    Protected Shared Function prepareStatement(ByVal sql As String, ByVal options() As String) As String
        Dim s As String = ""
        Dim i As Integer
        Dim n As Integer = 0

        If options Is Nothing Or UBound(options, 1) < 0 Then
            s = sql
        Else
            For i = 1 To Len(sql)
                If Mid(sql, i, 1) = "?" Then
                    If n <= UBound(options, 1) Then
                        s &= options(n)
                        n += 1
                    Else

                    End If
                Else
                    s &= Mid(sql, i, 1)
                End If
            Next
        End If
        Return s

    End Function

    Protected Function loadByPrimaryKey(ByVal key As String) As DataSet

        Dim options(1) As String
        options(0) = key
        Try
            Dim stmt As String = prepareStatement(loadStmt, options)
            Dim da As New OleDb.OleDbDataAdapter(stmt, CN.ConnectionString)
            Dim ds As New DataSet(DS_DATA_SET)
            da.Fill(ds, DS_DATA_SET)

            Return ds
        Catch ex As Exception
            MsgBox("Error while retrieving record from " & localTable & ": " & ex.Message)
            Return Nothing
        End Try

    End Function

    Protected Function loadAll() As DataSet

        Try
            Dim da As New OleDb.OleDbDataAdapter(loadAllStmt, CN.ConnectionString)
            Dim ds As New DataSet(DS_DATA_SET)
            da.Fill(ds, DS_DATA_SET)
            Return ds
        Catch ex As Exception
            MsgBox("Error while retrieving records from " & localTable & ": " & ex.Message)
            Return Nothing
        End Try

    End Function

    Protected Function load(ByVal stmt As String, ByVal options() As String) As DataSet

        If Not IsNothing(options) Then
            If Trim(options(0)) <> "" Then
                stmt = prepareStatement(stmt, options)
            End If
        End If
        Dim da As New OleDb.OleDbDataAdapter(stmt, CN.ConnectionString)
        Dim ds As New DataSet(DS_DATA_SET)
        da.Fill(ds, DS_DATA_SET)

        Return ds

    End Function

    Protected Function updateRecord(ByVal key As String, ByVal values() As String) As Boolean
        Try
            Dim cmd As New OleDb.OleDbCommand
            cmd.CommandType = CommandType.Text
            cmd.Connection = CN
            cmd.CommandText = prepareStatement(Me.updateStmt, values)
            cmd.ExecuteNonQuery()
        Catch ex As Exception
            MsgBox("Error while updating record: " & ex.Message)
            Return False
        End Try
        Return True
    End Function

    Protected Function insertRecord(ByVal values() As String) As Integer
        Dim id As Integer = -1
        Try
            Dim da As New OleDb.OleDbDataAdapter(loadAllStmt, CN.ConnectionString)
            Dim ds As New DataSet(DS_DATA_SET)
            da.Fill(ds, DS_DATA_SET)
            'Dim cmd As New OleDb.OleDbCommandBuilder(da)
            'cmd.QuotePrefix = "[" : cmd.QuoteSuffix = "]"
            'da.InsertCommand = cmd.getinsertcommand

            Dim cmd As New OleDb.OleDbCommand
            cmd.CommandType = CommandType.Text
            cmd.Connection = CN
            cmd.CommandText = prepareStatement(Me.insertStmt, values)
            'Dim parm As OleDb.OleDbParameter = New OleDb.OleDbParameter("@newID", OleDb.OleDbType.Integer, 4, priKeyName)
            'parm.Direction = ParameterDirection.Output
            'cmd.Parameters.Add(parm)
            cmd.ExecuteNonQuery()
            'id = CType(parm.Value, Integer)
            id = 0
        Catch ex As OleDb.OleDbException
            If InStr(LCase(ex.Message), "duplicate") > 0 Then
                MsgBox("Attempt to add a duplicate record.")
            Else
                MsgBox("Error while inserting record into " & localTable & ": " & ex.Message)
            End If
            Return id
        Catch ex As Exception
            MsgBox("Error while inserting record into " & localTable & ": " & ex.Message)
            Return id
        End Try
        Return id
    End Function

    Public Function delete(ByVal key As String) As Boolean

        Try
            Dim values() As String = {key}
            Dim cmd As New OleDb.OleDbCommand
            cmd.CommandType = CommandType.Text
            cmd.Connection = CN
            cmd.CommandText = prepareStatement(SQL_DELETE, values)
            cmd.ExecuteNonQuery()
        Catch ex As Exception
            MsgBox("Error while deleting record: " & ex.Message)
            Return False
        End Try
        Return True

    End Function
End Class
